src/bin/pgcopydb/pgsql_timeline.c

```diff
@@ -0,0 +1,363 @@
+		return;
+	}
+
+	/* filename (text) */
+	char *value = PQgetvalue(result, 0, 0);
+	strlcpy(context->filename, value, sizeof(context->filename));
+
+	/* content (bytea) */
+	value = PQgetvalue(result, 0, 1);
+
+	if (strlen(value) >= sizeof(context->content))
+	{
+		log_error("Received a timeline history file of %zu bytes, "
+				  "pgcopydb is limited to files of up to %zu bytes.",
+				  strlen(value),
+				  sizeof(context->content));
+		context->parsedOk = false;
+	}
+	strlcpy(context->content, value, sizeof(context->content));
+
+	context->parsedOk = true;
+}
+
+
+/*
+ * parseTimelineHistory parses the content of a timeline history file.
+ */
+bool
+parseTimelineHistory(const char *filename, const char *content,
+					 IdentifySystem *system, void *context)
+{
+	LinesBuffer lbuf = { 0 };
+
+	if (!splitLines(&lbuf, (char *) content))
+	{
+		/* errors have already been logged */
+		return false;
+	}
+
+	uint64_t prevend = InvalidXLogRecPtr;
+
+	TimelineHistoryEntry *entry = &system->currentTimeline;
+	int timelineCount = 0;
+
+	for (uint64_t lineNumber = 0; lineNumber < lbuf.count; lineNumber++)
+	{
+		char *ptr = lbuf.lines[lineNumber];
+
+		/* skip leading whitespace and check for # comment */
+		for (; *ptr; ptr++)
+		{
+			if (!isspace((unsigned char) *ptr))
+			{
+				break;
+			}
+		}
+
+		if (*ptr == '\0' || *ptr == '#')
+		{
+			continue;
+		}
+
+		log_trace("parseTimelineHistory line %lld is \"%s\"",
+				  (long long) lineNumber,
+				  lbuf.lines[lineNumber]);
+
+		char *tabptr = strchr(lbuf.lines[lineNumber], '\t');
+
+		if (tabptr == NULL)
+		{
+			log_error("Failed to parse history file line %lld: \"%s\"",
+					  (long long) lineNumber, ptr);
+			return false;
+		}
+
+		*tabptr = '\0';
+
+		if (!stringToUInt(lbuf.lines[lineNumber], &entry->tli))
+		{
+			log_error("Failed to parse history timeline \"%s\"", tabptr);
+			return false;
+		}
+
+		char *lsn = tabptr + 1;
+
+		for (char *lsnend = lsn; *lsnend; lsnend++)
+		{
+			if (!(isxdigit((unsigned char) *lsnend) || *lsnend == '/'))
+			{
+				*lsnend = '\0';
+				break;
+			}
+		}
+
+		if (!parseLSN(lsn, &entry->end))
+		{
+			log_error("Failed to parse history timeline %d LSN \"%s\"",
+					  entry->tli, lsn);
+			return false;
+		}
+
+		entry->begin = prevend;
+		prevend = entry->end;
+
+		log_trace("parseTimelineHistory[%d]: tli %d [%X/%X %X/%X]",
+				  timelineCount,
+				  entry->tli,
+				  LSN_FORMAT_ARGS(entry->begin),
+				  LSN_FORMAT_ARGS(entry->end));
+
+		catalog_add_timeline_history(context, entry);
+		timelineCount++;
+	}
+
+	/*
+	 * Create one more entry for the "tip" of the timeline, which has no entry
+	 * in the history file.
+	 */
+	entry->tli = system->timeline;
+	entry->begin = prevend;
+	entry->end = InvalidXLogRecPtr;
+
+	log_trace("parseTimelineHistory[%d]: tli %d [%X/%X %X/%X]",
+			  timelineCount,
+			  entry->tli,
+			  LSN_FORMAT_ARGS(entry->begin),
+			  LSN_FORMAT_ARGS(entry->end));
+
+	catalog_add_timeline_history(context, entry);
```

Comment:
We need to either add `(void) ` as a prefix to mean we don't care about the result, or rather properly handle the boolean value returned and error out when it's not true.

Reply:
Fixed.

This is not the first time I missed checking return values on a function. I wonder I should use `-Wunused-result` or set up a linter locally.